src/cargo/ops/cargo_install.rs

```diff
@@ -493,7 +534,137 @@ fn install_one(
                 ),
             )?;
         }
-        Ok(())
+        Ok(false)
+    }
+}
+
+fn is_installed(
+    pkg: &Package,
+    config: &Config,
+    opts: &ops::CompileOptions,
+    rustc: &Rustc,
+    target: &str,
+    root: &Filesystem,
+    dst: &Path,
+    force: bool,
+) -> CargoResult<bool> {
+    let tracker = InstallTracker::load(config, root)?;
+    let (freshness, _duplicates) =
+        tracker.check_upgrade(dst, pkg, force, opts, target, &rustc.verbose_version)?;
+    Ok(freshness == Freshness::Fresh)
+}
+
+/// Checks if vers can only be satisfied by exactly one version of a package in a registry, and it's
+/// already installed. If this is the case, we can skip interacting with a registry to check if
+/// newer versions may be installable, as no newer version can exist.
+fn installed_exact_package<T>(
+    dep: Dependency,
+    source: &mut T,
+    config: &Config,
+    opts: &ops::CompileOptions,
+    root: &Filesystem,
+    dst: &Path,
+    force: bool,
+) -> CargoResult<Option<Package>>
+where
+    T: Source,
+{
+    if !dep.version_req().is_exact() {
+        // If the version isn't exact, we may need to update the registry and look for a newer
+        // version - we can't know if the package is installed without doing so.
+        return Ok(None);
+    }
+    // Try getting the package from the registry  without updating it, to avoid a potentially
+    // expensive network call in the case that the package is already installed.
+    // If this fails, the caller will possibly do an index update and try again, this is just a
+    // best-effort check to see if we can avoid hitting the network.
+    if let Ok(pkg) = select_pkg(
+        source,
+        DependencyOrListAllFn::<fn(&mut T) -> CargoResult<Vec<Package>>>::Dependency(dep),
+        config,
+        false,
+    ) {
+        let (_ws, rustc, target) =
+            make_ws_rustc_target(&config, opts, &source.source_id(), pkg.clone())?;
+        if let Ok(true) = is_installed(&pkg, config, opts, &rustc, &target, root, &dst, force) {
+            return Ok(Some(pkg));
+        }
+    }
+    Ok(None)
+}
+
+fn make_ws_rustc_target<'cfg>(
+    config: &'cfg Config,
+    opts: &ops::CompileOptions,
+    source_id: &SourceId,
+    pkg: Package,
+) -> CargoResult<(Workspace<'cfg>, Rustc, String)> {
+    let mut ws = if source_id.is_git() {
+        // Don't use ws.current() in order to keep the package source as a git source so that
+        // install tracking uses the correct source.
```

Comment:
I think this comment is no longer valid, since in the old code it was referring to returning the Package.

I think it needs to be rewritten and moved to the `let git_package =` line above, to explain why the `git_package` local exists.